README.md
# SatNOGS Waterfall Tabulation Helper

Repository: [kerel-fs/satnogs-waterfall-tabulation-helper](https://gitlab.com/kerel-fs/satnogs-waterfall-tabulation-helper)

The SatNOGS Waterfall Tabulation helper is a script to extract rough doppler measurements
from SatNOGS waterfall png files. It downloads the waterfall png and obervation metadata and
allows the user to extract doppler measurements interactively.
Those measurements are stored in a STRF-compatible data file.

## Installation & Usage

Follow the  [Usage Guide](./docs/usage_guide.md).

## History

This tool evolved from the need to tabulate doppler data from waterfall images by DF2MZ of the lunar Change-4 probe,
see [DF2MZ Doppler Measurements](https://gitlab.com/kerel-fs/jupyter-notebooks/tree/master/change4/data#df2mz-doppler-measurements).
It was adjusted to read SatNOGS waterfall pngs instead and extended to
remove the doppler correction performed by the satnogs-client during the observation.
Its purpose is similar to rfplot in [STRF](https://github.com/cbassa/strf).

Original release post: [New software: SatNOGS waterfall tabulation helper](https://community.libre.space/t/new-software-satnogs-waterfall-tabulation-helper/4380)

From 2021-11-20 until 2023-01-13 it was maintained as part of [STRF](https://github.com/cbassa/strf).

While still supporting strf, 2023-01-13 it was moved into its own repository at
[kerel-fs/satnogs-waterfall-tabulation-helper](https://gitlab.com/kerel-fs/satnogs-waterfall-tabulation-helper)
to ease the development process.

## License

AGPL-3.0-or-later


CHANGELOG.md
# Changelog
## 0.3
- Fixed data path (correct path: `~/.local/share/satnogs-tabulation-helper/`)
- Fixed marker input (rounded to nearest integer pixel value).
- Added STRF rffit hint text (by using `--show-hint`).

## 0.2

First release,
code imported from [kerel-fs/strf@2efec5](https://gitlab.com/kerel-fs/strf/-/commit/2efec5a386dfe2888cc85dc3628a7812d33f4ffc).

- No preparation of data directories needed anymore. This will be done automatically.
  Data is stored in `~/.local/share/satnogs-tabulation-helper/` now.


settings.py
import sys
import os

from pathlib import Path
from decouple import config, UndefinedValueError

home_dir = os.path.expanduser('~')
data_dir = Path(home_dir, '.local/share/satnogs-tabulation-helper/')

try:
    TLE_DIR = config('SATNOGS_TLE_DIR', default=Path(data_dir, 'tles'))
    WATERFALL_IMAGES_DIR = config('SATNOGS_WATERFALL_IMAGES_DIR', default=Path(data_dir, 'waterfalls'))
    ARTIFACTS_DIR = config('SATNOGS_ARTIFACTS_DIR', default=Path(data_dir, 'artifacts'))
    DOPPLER_OBS_DIR = config('SATNOGS_DOPPLER_OBS_DIR', default=Path(data_dir, 'doppler_obs'))
    OBSERVATIONS_DIR = config('SATNOGS_OBSERVATIONS_DIR', default=Path(data_dir, 'observations'))
    SITES_TXT = config('SATNOGS_SITES_TXT', default=Path(data_dir, 'sites.txt'))


    SATNOGS_NETWORK_API_URL = config('SATNOGS_NETWORK_API_URL', default="https://network.satnogs.org/api/")
    SATNOGS_DB_API_URL = config('SATNOGS_DB_API_URL', default="https://db.satnogs.org/api/")
    SATNOGS_DB_API_TOKEN = config('SATNOGS_DB_API_TOKEN')
except UndefinedValueError as err:
    print("Missing Environment Variable.")
    print(err)
    sys.exit(-1)

class strf_path:
    def init_paths():
        # Make sure xdg-data-dir for this application exists
        data_dir.mkdir(exist_ok=True)
        for subdir in ['tles', 'waterfalls', 'artifacts', 'doppler_obs', 'observations']:
            Path(data_dir, subdir).mkdir(exist_ok=True)
        if not Path(data_dir, 'sites.txt').exists():
            Path(data_dir, 'sites.txt').touch()

    def artifact(observation_id):
        return '{}/{}.h5'.format(ARTIFACTS_DIR, observation_id)
    def doppler_obs(observation_id):
        return '{}/{}.dat'.format(DOPPLER_OBS_DIR, observation_id)
    def observation_tle(observation_id):
        return '{}/{}.txt'.format(TLE_DIR, observation_id)
    def observation(observation_id):
        return '{}/{}.json'.format(OBSERVATIONS_DIR, observation_id)
    def sites_txt():
        return SITES_TXT
    def waterfall_image(observation_id):
        return '{}/{}.png'.format(WATERFALL_IMAGES_DIR, observation_id)
    def data_dir():
        return data_dir


download_satnogs_tle.py
#!/usr/bin/env python3

import argparse
import os

from settings import strf_path
from satnogs_api_client import fetch_observation_data


def download_satnogs_tle(observation_id):
    filename = strf_path.observation_tle(observation_id)

    if os.path.exists(filename):
        # Skip
        print(f'TLE for {observation_id} exists already, skip.')
        return

    obs = fetch_observation_data([observation_id])[0]

    with open(filename, 'w') as f:
        f.write(obs['tle0'])
        f.write('\n')
        f.write(obs['tle1'])
        f.write('\n')
        f.write(obs['tle2'])
        f.write('\n')
    print("TLE saved in {}".format(filename))


if __name__ == '__main__':
    strf_path.init_paths()

    parser = argparse.ArgumentParser(description='Download TLE from a specific Observation in SatNOGS Network.')
    parser.add_argument('observation_id', type=int,
                        help='SatNOGS Observation ID')
    args = parser.parse_args()

    download_satnogs_tle(args.observation_id)


satnogs_waterfall_tabulation_helper.py
#!/usr/bin/env python3

import argparse
import csv
import ephem
import logging
import requests

from astropy.time import Time
from astropy import constants as const
from datetime import datetime, timedelta
from io import BytesIO
import matplotlib.pyplot as plt
import numpy as np
from os import path
from PIL import Image
from settings import strf_path, SITES_TXT

from satnogs_api_client import fetch_observation_data, fetch_tle_of_observation

logger = logging.getLogger(__name__)

__version__ = '0.3'

STRF_HINT_TEXT = """
Modify the following line:
ST_DATADIR="$HOME/src/strf" # path to your strf directory

Then run the following commands:
cd $ST_DATADIR

STH_DATA_DIR="{sth_data_dir}"
OBS_ID={observation_id}
SITE_ID={site_id}
NORAD_ID={norad_id}

ST_TLEDIR=$STL_DATA_DIR/tles
ST_COSPAR={site_id}

export ST_DATADIR
export ST_TLEDIR
export ST_COSPAR

./rffit -d $STH_DATA_DIR/doppler_obs/$OBS_ID.dat -i $NORAD_ID -c $STH_DATA_DIR/tles/$OBS_ID.txt -s $SITE_ID
"""


def tabulation_helper_dialog(lower_left, upper_right,
                             bandwidth, duration,
                             f_center, f_shift, filename_out,
                             waterfall_matrix,
                             epoch_start, site_id, correction_method=None):
    # Assumptions: x-axis - freqency, y-axis - time; up-time advance, right-freq advance
    # Derive conversion parameters
    x_range = upper_right[0] - lower_left[0]
    y_range = lower_left[1] - upper_right[1]
    x_offset = int(lower_left[0] + 0.5 * x_range)
    y_offset = lower_left[1]

    t_step = duration / y_range
    f_step = bandwidth / x_range

    # Highlight center line
    # waterfall_matrix[:,x_offset] = (255,0,0,255)

    # Highlight calibration lines
    waterfall_matrix[y_offset,:] = (255,0,0,255)
    waterfall_matrix[:,upper_right[0]] = (255,0,0,255)
    waterfall_matrix[upper_right[1],:] = (255,0,0,255)

    markers = {'x': [], 'y': [], 't': [], 'f': []}

    fig, ax = plt.subplots()
    ax.imshow(waterfall_matrix)
    marker_line, = ax.plot([0], [0], marker='.', c='k', zorder=100)

    def on_mouse_press(event):
        tb = plt.get_current_fig_manager().toolbar
        if not (event.button==1 and event.inaxes and tb.mode == ''):
            return

        event_x = round(event.xdata)
        event_y = round(event.ydata)

        markers['x'].append(event_x)
        markers['y'].append(event_y)
        markers['t'].append(epoch_start - (event_y - y_offset) * t_step)
        markers['f'].append(f_step * (event_x - x_offset) + f_shift)
        logger.info('x:{:.2f}  y:{:.2f}  --> {:7.0f} -->  {}   {:7.0f}'.format(markers['x'][-1],
                                                                                markers['y'][-1],
                                                                                markers['f'][-1] - f_center,
                                                                                markers['t'][-1],
                                                                                markers['f'][-1]))
        update_plot_markers()

    def update_plot_markers():
        marker_line.set_xdata(markers['x'])
        marker_line.set_ydata(markers['y'])
        # ax.scatter(x=[int(event.xdata)], y=[int(event.ydata)], marker='.', c='k', zorder=100)
        ax.figure.canvas.draw()

    def undo_track_point_selection():
        markers['x'] = markers['x'][:-1]
        markers['y'] = markers['y'][:-1]
        markers['t'] = markers['t'][:-1]
        markers['f'] = markers['f'][:-1]
        update_plot_markers()

    def write_strf_spectrum(filename, site_id, markers):
        with open(filename, 'w') as f:
            for t,freq in list(sorted(zip(markers['t'],markers['f']), key=lambda x: x[0])):

                if correction_method:
                    freq_recv = correction_method(t, freq)
                else:
                    freq_recv = freq
                line = '{:.6f}\t{:.2f}\t1.0\t{}\n'.format(Time(t).mjd, freq_recv, site_id)
                logger.debug(line, end='')
                f.write(line)

    def save_selected_track_points():
        write_strf_spectrum(filename_out, site_id, markers)
        logger.info('Stored {} selected track points in {}'.format(len(markers['t']), filename_out))

    def on_key(event):
        if event.key == 'u':
            undo_track_point_selection()
        elif event.key == 'f':
            save_selected_track_points()

    fig.canvas.mpl_connect('button_press_event', on_mouse_press)
    fig.canvas.mpl_connect('key_press_event', on_key)
    plt.tight_layout()
    plt.show()


def read_sitestxt(path):
    sites = {}
    with open(path) as fp:
        d = csv.DictReader([row for row in fp if not row.startswith('#')],
                           delimiter=' ',
                           fieldnames=["no", "id", "lat", "lon", "alt"],
                           restkey="observer",
                           skipinitialspace=True)
        for line in d:
            sites_id = int(line["no"])
            sites[sites_id] = {'lat': line["lat"],
                               'lon': line["lon"],
                               'sn': line["id"],
                               'alt': line["alt"],
                               'name': ' '.join(line["observer"])}
    return sites


def add_station_to_sitestxt(station):
    # Read the sites.txt
    sites = read_sitestxt(strf_path.sites_txt())
    if station['id'] in sites.keys():
        # Station already listed
        return
    else:
        # New station, add entry
        entry = '{id:04d} SN  {lat:9.5f} {lng:9.5f} {alt:5d}    {name}\n'.format(**station)

        with open(SITES_TXT, 'a') as f:
            f.write(entry)

def print_strf_hint(observation_id, site_id, norad_id):
    print(STRF_HINT_TEXT.format(
        sth_data_dir=strf_path.data_dir(),
        observation_id=observation_id,
        site_id=site_id,
        norad_id=norad_id
    ))

def tabulation_helper(observation_id, show_hint=False):
    # Fetch waterfall image and observation data
    observation = fetch_observation_data([observation_id])[0]
    tle = fetch_tle_of_observation(observation_id)

    if observation['vetted_status'] == 'failed':
        print(f'Skip failed observation {observation_id}')
        return

    result = requests.get(observation['waterfall'])
    image = Image.open(BytesIO(result.content))
    waterfall_matrix = np.array(image)

    epoch_start = datetime.strptime(observation['start'], '%Y-%m-%dT%H:%M:%SZ')
    epoch_end = datetime.strptime(observation['end'], '%Y-%m-%dT%H:%M:%SZ')
    site_id = int(str(observation['ground_station']))

    station = {'lat': observation['station_lat'],
               'lng': observation['station_lng'],
               'alt': observation['station_alt'],
               'name': "SatNOGS No.{}".format(observation['ground_station']),
                'id': observation['ground_station']}

    # TODO: Find a method to fetch this data from satnogs-db
    # NOTE: This is a very bad heuristic  and will fail in many cases!
    if observation['norad_cat_id'] == 46922:
        bandwidth = 230.4e3 # Hz
    else:
        bandwidth = 48e3 # Hz

    # Store TLE to file
    with open(strf_path.observation_tle(observation_id), 'w') as f:
        f.write('0 OBJECT\n')
        for line in tle:
            f.write(f'{line}\n')

    # Add SatNOGS station to sites.txt
    add_station_to_sitestxt(station)

    # Set image parameters
    lower_left = (66, 1553)
    upper_right = (686, 13)

    # Set data parameters
    duration = epoch_end - epoch_start
    f = observation['transmitter_downlink_low']
    drift = observation['transmitter_downlink_drift']
    if drift:
        f_center = f * (1 + drift / 1e9)
    else:
        f_center = f
    f_shift = 0 # -18.0 # kHz, when measuring left tone of bobcat-1 transmissions

    filename_out = strf_path.doppler_obs(observation_id)

    # Initialize SGP4 propagator / pyephem
    satellite = ephem.readtle('sat', tle[0], tle[1])
    observer = ephem.Observer()
    observer.lat = str(station['lat'])
    observer.lon = str(station['lng'])
    observer.elevation = station['alt']

    # Extract norad id from TLE (can be different to the norad id of the Satellite entry in SatNOGS DB)
    norad_id_from_tle = satellite.catalog_number
    if norad_id_from_tle != observation['norad_cat_id']:
        print('WARNING: NORAD ID mismatch between Observation Metadata and TLE '
              f'({observation["norad_cat_id"]} vs {norad_id_from_tle})!\n'
              'Use NORAD ID from TLE for rffit.'
              )

    def remove_doppler_correction(t, freq):
        # Remove the doppler correction
        observer.date = t
        satellite.compute(observer)
        v = satellite.range_velocity
        df = f_center * v / const.c.value
        return  f_center + freq - df

    tabulation_helper_dialog(lower_left, upper_right,
                             bandwidth, duration,
                             f_center, f_shift, filename_out,
                             waterfall_matrix,
                             epoch_start, site_id, correction_method=remove_doppler_correction)
    if show_hint:
        print_strf_hint(observation_id, site_id, norad_id_from_tle)

if __name__ == '__main__':
    strf_path.init_paths()

    parser = argparse.ArgumentParser(description='Interactive helper to tabulate signals from SatNOGS waterfalls.')
    parser.add_argument('observation_ids', metavar='ID', type=int, nargs='+',
                        help='SatNOGS Observation ID')
    parser.add_argument(
        "--version",
        action="version",
        version=f'SatNOGS Waterfall Tabulation Helper version {__version__}',
    )
    parser.add_argument(
        "--show-hint",
        action="store_true",
        help='Output hints for using strf rffit',
    )

    args = parser.parse_args()

    logging.basicConfig(level=logging.INFO,
                        format="%(levelname)s: %(message)s")

    for observation_id in args.observation_ids:
        if path.exists(strf_path.doppler_obs(observation_id)):
            logger.warning(f'Observation {observation_id} skipped, previous analysis found.\n'
                           f'Remove the following file if you want to re-run the analysis:\n'
                           f'{strf_path.doppler_obs(observation_id)}\n')
            continue
        tabulation_helper(observation_id, args.show_hint)
